### ISSUE TITLE
Return type of actions

### ISSUE THREAD

**Issue by corydeppen** [xdave/typescript-fsa-redux-thunk#9]
I'm using `typescript-fsa-redux-thunk@2.0.0-beta.6` and trying to understand the proper way to dispatch an action in a view. Pre-v2 I was able to `dispatch(action()).then(res => /* Do something with the result */)`, but `bindThunkAction` returns a different type of promise now (e.g. `Promise<Action<Success<FormData, Case>>>`). As a result, it looks like I need to either explicitly assert the type of `dispatch` or include `isPromise`, `isSuccess`, and `isFailure` guards in my view.

Part of what confuses me is that it seems like there should only be two types that could be returned by `bindThunkAction` - success and failure. I don't understand why the return type can be something other than the action creator result and error types.

Am I misunderstanding the intended way to dispatch and use the promise?

**Comment by xdave** [comment]
So, the 2.0 beta version brought in a few design changes that obviously break the previous 1.0 behavior:

1. The return type doesn't have to be a promise
2. Errors are not re-thrown if caught inside of the worker

Because the error isn't being rethrown, the return-type of the thunk is ambiguous.

I'm considering creating an option to revert to the old 1.0 behavior.

So, if I am using a promise, the way I use it now:

```ts
const doThingAsync = create.async<State, undefined, void, Error>('DO_THING');
const doThing = bindThunkAction(doThingAsync, async () => {
    // noop
});

// ...

export const someFunc = async () => {
    const action = await store.dispatch(doThing());
    if (isSuccess(action)) {
        // Do something with result
    }
};
```

Apart from resolving the promise, using `await` on the result does the same thing as wrapping it in:
```ts
if (isPromise(action)) {
    // action has been asserted be a Promise
}
```

This issue hasn't come up for me because in my React+Redux projects, I generally pass in my actions to a call to `connect()` using `react-redux`.  Using an object literal to carry the actions has the added benefit of automatically getting `bindActionCreators` called on it.

I generally only end up dealing with the direct result of `dispatch()` when dealing with non-async/non-thunks.  The result is handled by the reducer and should update the state accordingly.  This state can than be pulled-in by your components.  For example:

```ts
import React = require('react');
import ReactDOM = require('react-dom');
import { connect, Provider } from 'react-redux';
import {
    createStore,
    combineReducers,
    applyMiddleware,
    Reducer
} from 'redux';
import thunk from 'redux-thunk';
import actionCreatorsFactory from 'typescript-fsa';
import bindThunkAction, { thunkToAction } from 'typescript-fsa-redux-thunk';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

interface State {
    stuff: {
        msg: string;
        loading?: boolean;
        error?: Error;
    };
}

const create = actionCreatorsFactory('things');

const getMsgAsync = create.async<State, undefined, string, Error>('GET_MSG');
const getMsg = bindThunkAction(getMsgAsync, () =>
    new Promise<string>(resolve => setTimeout(resolve, 1000, 'Hello')));

const initial: State['stuff'] = { msg: '' };

const stuff = reducerWithInitialState(initial)
    .case(getMsgAsync.started, state => ({
        ...state,
        ...initial,
        loading: true
    }))
    .case(getMsgAsync.failed, (state, { error }) => ({
        ...state,
        loading: false,
        error
    }))
    .case(getMsgAsync.done, (state, { result }) => ({
        ...state,
        msg: result,
        loading: false
    }))
    .build();

const reducer: Reducer<State> = combineReducers({
    stuff
});

const store = createStore(reducer, applyMiddleware(thunk));

const mapState = (state: State) => state;
const mapActions = {
    getMsg: thunkToAction(getMsg)
};

const C = connect(mapState, mapActions, (state, actions, props) => ({
    state: {
        loading: state.stuff.loading,
        msg: state.stuff.msg,
        error: state.stuff.error
    },
    event: {
        getMsg: () => actions.getMsg()
    },
    ...props
}))

const SomeComponent = C(({ state, event }) => (
    <React.Fragment>
        {state.loading && `Loading...`}
        {state.error && state.error.message}
        <br />
        <button onClick={event.getMsg} disabled={state.loading||!!state.error}>
            Clickity {state.msg}
        </button>
    </React.Fragment>
));

SomeComponent.displayName = 'SomeComponent';

const Index: React.SFC = () => (
    <Provider store={store}>
        <SomeComponent />
    </Provider>
);

ReactDOM.hydrate(<Index />, document.querySelector('body'));
```

**Comment by xdave** [comment]
@corydeppen So, my latest refactorings (not pushed yet) are making me go back to the 1.x behavior.  I'm blocked by what I think might be a bug in the new `typescript-fsa` 3.0 beta.  I'll keep you posted.

**Comment by xdave** [comment]
@corydeppen OK I've found a workaround for the `typescript-fsa` bug I was talking about, and I've published a new beta (2.0.0-beta.8).  Please check it out and let me know if you have any issues.
